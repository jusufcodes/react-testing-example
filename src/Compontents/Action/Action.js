import React, {useState} from "react";
import ShareIcon from "../Icon/Share.svg";
import FavoriteIcon from "../Icon/Favorite.svg";
import classNames from "classnames";

export const Action = (props) => {
  const [clickCount, setClickCount] = useState(0);

  const handleClick = (event) => {
    setClickCount(clickCount + 1);

    if (props.onClick) {
      props.onClick(event);
    }
  };

  return props.children ? (
    <button onClick={handleClick}>
      {props.children} ({clickCount})
    </button>
  ) : null;
};

export async function likeAndShare(postId) {
  const apiUrl = postId ? "posts" : "error";

  const response = await fetch(`https://jsonplaceholder.typicode.com/${apiUrl}`, {
    method: 'POST',
    body: JSON.stringify({
      id: postId
    }),
    headers: {
      "Content-type": "application/json; charset=UTF-8"
    }
  });

  const json = await response.json();

  return json.id;
}

export const FavoriteAction = (props) => {
  const [liked, setLike] = useState(false);

  async function handleClick() {
    const currentLiked = liked;

    setLike(!currentLiked);

    if (!props.postId) {
      return
    }

    const id = await likeAndShare(props.postId);

    if (typeof id === "number") {
      setLike(!currentLiked);
    } else {
      alert("error");
    }
  }

  return props.postId ? (
    <Action onClick={handleClick}>
      <img src={FavoriteIcon} alt="" className={classNames("favoriteIcon", {
        active: liked
      })}/>
    </Action>
  ) : null;
};

export const ShareAction = (props) => {
  const [shared, setShared] = useState(false);

  async function handleClick() {
    const currentShared = shared;

    setShared(!currentShared);

    if (!props.postId) {
      return
    }

    const id = await likeAndShare(props.postId);

    if (typeof id === "number") {
      setShared(!currentShared);
    } else {
      alert("error");
    }
  }

  return props.postId ? (
    <Action onClick={handleClick}>
      <img src={ShareIcon} alt="" className={classNames("shareIcon", {
        active: shared
      })}/>
    </Action>
  ) : null;
};