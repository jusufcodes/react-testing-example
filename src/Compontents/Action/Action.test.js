import React from 'react';
import { shallow } from 'enzyme';
import renderer from "react-test-renderer";
import {Action, FavoriteAction, likeAndShare, ShareAction} from "./Action";
import FavoriteIcon from "../Icon/Favorite.svg";
import ShareIcon from "../Icon/Share.svg";

const demoPosts = "https://jsonplaceholder.typicode.com/posts";

describe("Card action component", () => {
  /**
   * TEST REST API
   */
  test("Rest API", async () => {
    // FAIL
    const likedOrShared_error = await likeAndShare();
    expect(likedOrShared_error).toEqual(undefined);

    // SUCCESS
    const likedOrShared_success = await likeAndShare(101);
    expect(likedOrShared_success).toEqual(101);
  });

  /**
   * DEFAULT ACTION
   *  **/
  test("Action without text", () => {
    const action = <Action />;

    expect(shallow(action).isEmptyRender()).toEqual(true);

    const contentSnapshot = renderer.create(action).toJSON();
    expect(contentSnapshot).toMatchSnapshot();
  });

  test("Action with text", () => {
    const children = "Action 1";

    const action = <Action>{children}</Action>;

    expect(shallow(action).text()).toContain(children);

    const contentSnapshot = renderer.create(action).toJSON();
    expect(contentSnapshot).toMatchSnapshot();
  });

  test("Action onClick event", () => {
    const children = "Action 1";

    const handleClick = jest.fn();
    const action = shallow(<Action onClick={handleClick}>{children}</Action>);

    action.simulate("click");
    expect(handleClick.mock.calls.length).toEqual(1);

    action.simulate("click");
    expect(handleClick.mock.calls.length).toEqual(2);
  });

  /**
   *  FAVORITE
   *  **/
  test("Action as favorite - no props", () => {
    const action = <FavoriteAction />;

    expect(shallow(action).isEmptyRender()).toEqual(true);
  });

  test("Action as favorite - with PostID and Like", () => {
    const action = <FavoriteAction postId={1} />;
    const shallowAction = shallow(action);

    expect(shallowAction.find("img")).toHaveLength(1);
    expect(shallowAction.find("img").prop("src")).toEqual(FavoriteIcon);

    const contentSnapshot = renderer.create(action).toJSON();
    expect(contentSnapshot).toMatchSnapshot();

    shallowAction.simulate("click");
    shallowAction.update();

    expect(shallowAction.find("img").hasClass("active")).toBe(true);

    shallowAction.simulate("click");
    shallowAction.update();

    expect(shallowAction.find("img").hasClass("active")).toBe(false);
  });

  /**
   * SHARE
   * **/
  test("Action as share - no props", () => {
    const action = <ShareAction />;

    expect(shallow(action).isEmptyRender()).toEqual(true);
  });

  test("Action as share - with PostID and Shared", () => {
    const action = <ShareAction postId={1} />;
    const shallowAction = shallow(action);

    expect(shallowAction.find("img")).toHaveLength(1);
    expect(shallowAction.find("img").prop("src")).toEqual(ShareIcon);

    const contentSnapshot = renderer.create(action).toJSON();
    expect(contentSnapshot).toMatchSnapshot();

    shallowAction.simulate("click");
    shallowAction.update();

    expect(shallowAction.find("img").hasClass("active")).toBe(true);

    shallowAction.simulate("click");
    shallowAction.update();

    expect(shallowAction.find("img").hasClass("active")).toBe(false);
  });

});