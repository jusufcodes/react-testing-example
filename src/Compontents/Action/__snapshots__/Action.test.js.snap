// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card action component Action as favorite - with PostID and Like 1`] = `
<button
  onClick={[Function]}
>
  <img
    alt=""
    className="favoriteIcon"
    src="Favorite.svg"
  />
   (
  0
  )
</button>
`;

exports[`Card action component Action as share - with PostID and Shared 1`] = `
<button
  onClick={[Function]}
>
  <img
    alt=""
    className="shareIcon"
    src="Share.svg"
  />
   (
  0
  )
</button>
`;

exports[`Card action component Action with text 1`] = `
<button
  onClick={[Function]}
>
  Action 1
   (
  0
  )
</button>
`;

exports[`Card action component Action without text 1`] = `null`;
